File: app/src/main/java/com/adsamcik/signalcollector/data/MapLayer.java
package com.adsamcik.signalcollector.data;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.vimeo.stag.UseStag;

import java.util.ArrayList;

@UseStag
public class MapLayer {
	private String name;
	private ArrayList<ValueColor> values;

	private float top;
	private float right;
	private float bottom;
	private float left;

	public static final float MIN_LATITUDE = -90;
	public static final float MAX_LATITUDE = 90;
	public static final float MIN_LONGITUDE = -180;
	public static final float MAX_LONGITUDE = 180;

	MapLayer() {}

	public MapLayer(@NonNull String name, float top, float right, float bottom, float left) {
		this.name = name;
		this.top = top;
		this.right = right;
		this.bottom = bottom;
		this.left = left;
	}

	public static String[] toStringArray(MapLayer[] layerArray) {
		String[] array = new String[layerArray.length];
		for (int i = 0; i < layerArray.length; i++)
			array[i] = layerArray[i].name;
		return array;
	}

	public static int indexOf(MapLayer[] layerArray, String name) {
		for (int i = 0; i < layerArray.length; i++) {
			if (layerArray[i].name.equals(name))
				return i;
		}
		return -1;
	}

	public static boolean contains(MapLayer[] layerArray, String name) {
		for (MapLayer aLayerArray : layerArray) {
			if (aLayerArray.name.equals(name)) {
				return true;
			}
		}
		return false;
	}

	public class ValueColor {
		public final int color;
		public final String name;

		public ValueColor(String name, int color) {
			this.name = name;
			this.color = color;
		}
	}

	//GETTERS
	public @NonNull String getName() {
		return name;
	}

	public @Nullable ArrayList<ValueColor> getValues() {
		return values;
	}

	public float getTop() {
		return top;
	}

	public float getRight() {
		return right;
	}

	public float getBottom() {
		return bottom;
	}

	public float getLeft() {
		return left;
	}

	//SETTERS for STAG
	void setName(@NonNull String name) {
		this.name = name;
	}

	void setValues(@Nullable ArrayList<ValueColor> values) {
		this.values = values;
	}

	void setTop(float top) {
		this.top = top;
	}

	void setRight(float right) {
		this.right = right;
	}
	void setBottom(float bottom) {
		this.bottom = bottom;
	}

	void setLeft(float left) {
		this.left = left;
	}
}

File: app/src/main/java/com/adsamcik/signalcollector/data/MapLayer.kt
package com.adsamcik.signalcollector.data

import android.content.Context
import com.adsamcik.signalcollector.utility.CoordinateBounds
import com.adsamcik.signalcollector.utility.Preferences
import com.squareup.moshi.JsonClass

/**
 * Data class containing information about the name and boundaries.
 * Does not use [com.adsamcik.signalcollector.utility.CoordinateBounds] because Stag does not support this level of customization for TypeAdapters and custom type adapter is not a priority right now.
 */
//todo Update to use CoordinateBounds
@JsonClass(generateAdapter = true)
data class MapLayer(var name: String,
                    val bounds: CoordinateBounds,
                    /**
                     * Contains information for the legend
                     */
                    var values: Array<ValueColor>? = null) {

	constructor(name: String,
	            top: Double = MAX_LATITUDE,
	            right: Double = MAX_LONGITUDE,
	            bottom: Double = MIN_LATITUDE,
	            left: Double = MIN_LONGITUDE,
	            /**
	             * Contains information for the legend
	             */
	            values: Array<ValueColor>? = null) : this(name, CoordinateBounds(top, right, bottom, left), values)

	companion object {
		const val MIN_LATITUDE = -90.0
		const val MAX_LATITUDE = 90.0
		const val MIN_LONGITUDE = -180.0
		const val MAX_LONGITUDE = 180.0
		/**
		 * Checks if MapLayer is in given array
		 */
		fun contains(layerArray: Array<MapLayer>, name: String): Boolean =
				layerArray.any { it.name == name }

		fun mockArray(): Array<MapLayer> = arrayOf(
				MapLayer("Mock", 30.0, 30.0, -30.0, -30.0),
				MapLayer("Wifi", 0.0, 30.0, -20.0, -30.0),
				MapLayer("Cell", 30.0, 30.0, -30.0, -30.0),
				MapLayer("Cell", 60.0, 30.0, -30.0, -30.0),
				MapLayer("Cell", 30.0, 30.0, -30.0, -30.0)
		)
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (javaClass != other?.javaClass) return false

		other as MapLayer

		if (name != other.name) return false
		if (bounds != other.bounds) return false
		if (values != null) {
			if (other.values == null) return false
			if (!values!!.contentEquals(other.values!!)) return false
		} else if (other.values != null) return false

		return true
	}

	override fun hashCode(): Int {
		var result = name.hashCode()
		result = 31 * result + bounds.hashCode()
		result = 31 * result + (values?.contentHashCode() ?: 0)
		return result
	}
}

enum class LayerType {
	Location,
	Cell,
	WiFi;

	companion object {
		fun valueOfCaseInsensitive(value: String) = when (value.toLowerCase()) {
			Location.name.toLowerCase() -> Location
			Cell.name.toLowerCase() -> Cell
			WiFi.name.toLowerCase(), "wi-fi" -> WiFi
			else -> throw IllegalArgumentException("Value '$value' is not a valid layer type.")
		}

		fun fromPreference(context: Context, key: String, default: LayerType): LayerType {
			val stringName = Preferences.getPref(context).getString(key, null) ?: return default
			return valueOfCaseInsensitive(stringName)
		}
	}
}

@JsonClass(generateAdapter = true)
data class ValueColor(val name: String, val color: Int)
